### PR TITLE
Added "Porting PyTorch model to JAX" tutorial

### PR DESCRIPTION
**Pull request by vfdev-5**

### CODE REVIEW

File: docs/JAX_porting_PyTorch_model.md [hunk 1]
```diff
@@ -0,0 +1,1607 @@
...
+- StochasticDepth
+- MBConv
+- RelativePositionalMultiHeadAttention
+- SwapAxes
+- WindowPartition
+- WindowDepartition
+- PartitionAttentionLayer
+- MaxVitLayer
+- MaxVitBlock
+- MaxViT
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:131]:
```suggestion
- `Conv2dNormActivation`
- `SqueezeExcitation`
- `StochasticDepth`
- `MBConv`
- `RelativePositionalMultiHeadAttention`
- `SwapAxes`
- `WindowPartition`
- `WindowDepartition`
- `PartitionAttentionLayer`
- `MaxVitLayer`
- `MaxVitBlock`
- `MaxViT`
```

For consistency.

I also wonder if listing these items here with no context here is helpful, but I think the alternative (giving a short description of each item) might be too much information at this point, seeing as you will detail the implementations below.

File: docs/JAX_porting_PyTorch_model.md [hunk 2]
```diff
@@ -0,0 +1,1607 @@
...
+
+Flax NNX module is very similar to PyTorch `torch.nn` module and we can simply map the following modules between PyTorch and Flax:
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:133]:
```suggestion
The Flax NNX module is very similar to PyTorch `torch.nn` module and we can map the following modules between PyTorch and Flax:
```

File: docs/JAX_porting_PyTorch_model.md [hunk 3]
```diff
@@ -0,0 +1,1607 @@
...
+- `nn.Linear` -> `nnx.Linear`
+- `nn.Conv2d` -> `nnx.Conv`
+- `nn.BatchNorm2d` -> `nnx.BatchNorm`
+- Activations like `nn.ReLU` -> `lambda x: nnx.relu(x)`
+- Pooling layers like `nn.AvgPool2d` -> `lambda x: nnx.avg_pool(x, ...)`
+- `nn.AdaptiveAvgPool2d(1)` -> `lambda x: nnx.avg_pool(x, (x.shape[1], x.shape[2]))`, x is in NHWC format
+- `nn.Flatten()` -> `lambda x: x.reshape(x.shape[0], -1)`
+
+
+If PyTorch model defines a learnable parameter and a buffer:
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:144]:
```suggestion
If the PyTorch model defines a learnable parameter and a buffer:
```

File: docs/JAX_porting_PyTorch_model.md [hunk 4]
```diff
@@ -0,0 +1,1607 @@
...
+```python
+class Model(nn.Module):
+    def __init__(self, ...):
+        ...
+        self.p = nn.Parameter(torch.ones(10))
+        self.register_buffer("b", torch.ones(5))
+```
+equivalent code in Flax would be
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:152]:
```suggestion
the equivalent code in Flax would be
```

File: docs/JAX_porting_PyTorch_model.md [hunk 5]
```diff
@@ -0,0 +1,1607 @@
...
+        v.value.mean() if v.value is not None else None
+    )
+
+for k, v in nnx.state(nnx_module, (nnx.BatchStat, Buffer)).flat_state().items():
+    print(
+        k,
+        v.value.mean() if v.value.dtype == "float32" else v.value.sum()
+    )
+```
+Equivalent PyTorch code is:
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:180]:
```suggestion
The equivalent PyTorch code is:
```

File: docs/JAX_porting_PyTorch_model.md [hunk 6]
```diff
@@ -0,0 +1,1607 @@
...
+
++++
+
+Please note the some tiny differences between `torch.nn` and Flax when porting models:
+- We should pass `rngs` to all NNX modules with parameters: e.g. `nnx.Linear(..., rngs=nnx.Rngs(0))`
+- Convolution 2D in Flax:
+  - we need explicitly define `kernel_size`, `strides` as two ints tuples, e.g. `(3, 3)`
+  - if `padding` should be defined as integer, e.g. 2, it should be defined as `((2, 2), (2, 2))`
+- Batch Normalization: momentum in `torch.nn` is defined as `1.0 - momentum` in Flax
+- 4D input array in Flax should be in NHWC format, i.e. of shape (N, H, W, C) compared to NCHW format (or (N, C, H, W) shape) in PyTorch.
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:202]:
```suggestion
Please note there are some differences between `torch.nn` and Flax when porting models:
- We should pass `rngs` to all NNX modules with parameters: e.g. `nnx.Linear(..., rngs=nnx.Rngs(0))`
- For a 2d convolution:
  - In Flax, we need to explicitly define `kernel_size`, `strides` as two ints tuples, e.g. `(3, 3)`
  - If `padding` should be defined as integer in PyTorch, e.g. 2, it should be defined as `((2, 2), (2, 2))` in Flax.
- For a batch normalization: the momentum in `torch.nn` is defined as `1.0 - momentum` in Flax
- 4D input arrays in Flax should be in NHWC format, i.e. of shape (N, H, W, C) compared to NCHW format (or (N, C, H, W) shape) in PyTorch.
```

Please make sure these are correct especially the padding item, the phrasing is what I am focusing on.

File: docs/JAX_porting_PyTorch_model.md [hunk 7]
```diff
@@ -0,0 +1,1607 @@
...
++++
+
+Below we implement one by one all the modules from the above list and add a simple forward pass check.
+
+```{code-cell} ipython3
+# Let's first implement equivalent of nn.Identity
+class Identity(nnx.Module):
+    def __call__(self, x: jax.Array) -> jax.Array:
+        return x
+```
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:213]:
````suggestion
Let's first implement equivalent of `nn.Identity`:

```{code-cell} ipython3
class Identity(nnx.Module):
    def __call__(self, x: jax.Array) -> jax.Array:
        return x
```
````

File: docs/JAX_porting_PyTorch_model.md [hunk 8]
```diff
@@ -0,0 +1,1607 @@
...
+    p_stochastic=[0.0, 0.2],
+)
+
+y = mod(x)
+print(y.shape)
+```
+
+#### `MaxVit` implementation
+
+Finally, we can assemble evreything together and define MaxVit model.
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:948]:
```suggestion
Finally, we can assemble everything together and define the MaxVit model.
```

File: docs/JAX_porting_PyTorch_model.md [hunk 9]
```diff
@@ -0,0 +1,1607 @@
...
+        jnp.abs(jax_output - jax_expected).mean(),
+    )
+
+
+t2f = Torch2Flax()
+```
+
+Let us now test our JAX modules. We only test the result of the forward pass in the inference mode such that we avoid disrcrepancies related to random layers like Dropout, StochasticDepth, etc.
+By default, we use absolute error tolerence `1e-3` when comparing the JAX output vs expected PyTorch result.
+For larger modules we set the device to CPU for JAX model to execute on in order to reduce the errors between CPU and CUDA.
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:1355]: (I think mode is meant to be model, but maybe I am wrong?)
```suggestion
Let us now test our JAX modules. We only test the result of the forward pass in the inference model such that we avoid discrepancies related to random layers like `Dropout`, `StochasticDepth`, etc.
By default, we use absolute error tolerence `1e-3` when comparing the JAX output vs expected PyTorch result.
For larger modules we set the device to CPU for the JAX model to execute on in order to reduce the errors between CPU and CUDA.
```

File: docs/JAX_porting_PyTorch_model.md [hunk 10]
```diff
@@ -0,0 +1,1607 @@
...
+)
+
+t2f.copy_module(torch_module, nnx_module, skip_modules=[
+    "WindowPartition", "WindowDepartition", "SwapAxes", "StochasticDepth",
+])
+
+test_modules(nnx_module, torch_module, torch.randn(4, 64, 56, 56), device="cpu")
+```
+
+Finally, we can check MaxVit implementation. Note that we raised absolute tolerence to `1e-1` when compare JAX output logits vs PyTorch expected logits.
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:1502]:
```suggestion
Finally, we can check the MaxVit implementation. Note that we raised the absolute tolerence to `1e-1` when comparing JAX output logits vs PyTorch expected logits.
```

File: docs/JAX_porting_PyTorch_model.md [hunk 11]
```diff
@@ -0,0 +1,1607 @@
...
+        self.max_seq_len = max_seq_len
+
+        self.to_qkv = nnx.Linear(feat_dim, self.n_heads * self.head_dim * 3, rngs=rngs)
+        self.scale_factor = feat_dim**-0.5
+
+        self.merge = nnx.Linear(self.head_dim * self.n_heads, feat_dim, rngs=rngs)
+
+        self.relative_position_index = Buffer(_get_relative_position_index(self.size, self.size))
+
+        # initialize with truncated normal the bias
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:539]:
```suggestion
        # initialize with truncated normal bias
```

File: docs/JAX_porting_PyTorch_model.md [hunk 12]
```diff
@@ -0,0 +1,1607 @@
...
+        return x
+
+
+class WindowDepartition(nnx.Module):
+    def __call__(self, x: jax.Array, p: int, h_partitions: int, w_partitions: int) -> jax.Array:
+        # Output array with expected layout of [B, H, W, C].
+        B, G, PP, C = x.shape
+        P = p
+        HP, WP = h_partitions, w_partitions
+        # split P * P dimension into 2 P tile dimensionsa
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:617]:
```suggestion
        # split P * P dimension into 2 P tile dimensions
```

File: docs/JAX_porting_PyTorch_model.md [hunk 13]
```diff
@@ -0,0 +1,1607 @@
...
+        # because we need to know hamy relative offsets there are in the grid
+        grid_size: Tuple[int, int],
+        mlp_ratio: int,
+        activation_layer: Callable,
+        norm_layer: Callable[..., nnx.Module],
+        attention_dropout: float,
+        mlp_dropout: float,
+        p_stochastic_dropout: float,
+        rngs: nnx.Rngs = nnx.Rngs(0),
+    ):
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:668]:
```suggestion
class PartitionAttentionLayer(nnx.Module):
    def __init__(
        self,
        in_channels: int,
        head_dim: int,
        # partitioning parameters
        partition_size: int,
        partition_type: str,
        # grid size needs to be known at initialization time
        # because we need to know hamy relative offsets there are in the grid
        grid_size: Tuple[int, int],
        mlp_ratio: int,
        activation_layer: Callable,
        norm_layer: Callable[..., nnx.Module],
        attention_dropout: float,
        mlp_dropout: float,
        p_stochastic_dropout: float,
        rngs: nnx.Rngs = nnx.Rngs(0),
    ):
    """Layer for partitioning the input tensor into non-overlapping windows and applying attention to each window.
    """
```

File: docs/JAX_porting_PyTorch_model.md [hunk 14]
```diff
@@ -0,0 +1,1607 @@
...
+        # account for stem stride
+        input_size = _get_conv_output_shape(input_size, kernel_size=3, stride=2, padding=1)
+        self.partition_size = partition_size
+
+        # blocks
+        blocks = []
+        in_channels = [stem_channels] + block_channels[:-1]
+        out_channels = block_channels
+
+        # precompute the stochastich depth probabilities from 0 to stochastic_depth_prob
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:1048]:
```suggestion
        # precompute the stochastic depth probabilities from 0 to stochastic_depth_prob
```

File: docs/JAX_porting_PyTorch_model.md [hunk 15]
```diff
@@ -0,0 +1,1607 @@
...
+plt.title(f"{labels[str(torch_class_id)]}\nScore: {torch_output.softmax(dim=-1)[0, class_id]:.4f}")
+plt.imshow(dog1.permute(1, 2, 0))
+
+plt.subplot(122)
+plt.title(f"{labels[str(flax_class_id)]}\nScore: {jax.nn.softmax(flax_output, axis=1)[0, flax_class_id]:.4f}")
+plt.imshow(dog1.permute(1, 2, 0))
+```
+
+```{code-cell} ipython3
+# Let's compute cosine distance between the logits:
```
Comment (melissawm) [inline docs/JAX_porting_PyTorch_model.md:1597]:
```suggestion

Let's compute cosine distance between the logits:

```{code-cell} ipython3
```

**Review by melissawm — COMMENTED** [review]
Hi @vfdev-5 - found a few typos, left a few suggestions for rewording and for adding links. Some are very optional, feel free to take or leave 😄 Cheers!

File: docs/JAX_porting_PyTorch_model.md [hunk 16]
```diff
@@ -0,0 +1,1607 @@
...
+- MaxVitLayer
+- MaxVitBlock
+- MaxViT
+
+Flax NNX module is very similar to PyTorch `torch.nn` module and we can simply map the following modules between PyTorch and Flax:
+- `nn.Sequential` and `nn.ModuleList` -> `nnx.Sequential`
+- `nn.Linear` -> `nnx.Linear`
+- `nn.Conv2d` -> `nnx.Conv`
+- `nn.BatchNorm2d` -> `nnx.BatchNorm`
+- Activations like `nn.ReLU` -> `lambda x: nnx.relu(x)`
```
Comment (jakevdp) [inline docs/JAX_porting_PyTorch_model.md:138]: No need to wrap in a `lambda` here:
```suggestion
- Activations like `nn.ReLU` -> `nnx.relu`
```

File: docs/JAX_porting_PyTorch_model.md [hunk 17]
```diff
@@ -0,0 +1,1607 @@
...
+- Pooling layers like `nn.AvgPool2d` -> `lambda x: nnx.avg_pool(x, ...)`
```
Comment (jakevdp) [inline docs/JAX_porting_PyTorch_model.md:139]: Maybe the same here? I'm not sure what `...` means.

File: docs/JAX_porting_PyTorch_model.md [hunk 18]
```diff
@@ -0,0 +1,1607 @@
...
+- `nn.AdaptiveAvgPool2d(1)` -> `lambda x: nnx.avg_pool(x, (x.shape[1], x.shape[2]))`, x is in NHWC format
+- `nn.Flatten()` -> `lambda x: x.reshape(x.shape[0], -1)`
```
Comment (jakevdp) [inline docs/JAX_porting_PyTorch_model.md:141]: maybe use `jax.vmap(jnp.ravel)` here? Or is that too magic?

**Review by jakevdp — CHANGES_REQUESTED** [review]
This is really great – thanks! Just a couple small comments below, then I'd love to get this merged!

File: docs/JAX_porting_PyTorch_model.md [hunk 19]
```diff
@@ -0,0 +1,1607 @@
...
+```{code-cell} ipython3
+import torch
+
+
+torch_model.eval()
+with torch.inference_mode():
+    x = torch.rand(2, 3, 224, 224)
+    output = torch_model(x)
+
+print(output.shape)
```
[the remark this answers is not in the document]
Comment (vfdev-5) [reply docs/JAX_porting_PyTorch_model.md:78]: It shows the shape of logits: e.g (2, 1000)

File: docs/JAX_porting_PyTorch_model.md (hunk 1 above)
Comment (vfdev-5) [reply docs/JAX_porting_PyTorch_model.md:131]: I can rearrange them in the following way:
```
To port the [PyTorch implementation of the MaxVit model](https://github.com/pytorch/vision/blob/945bdad7523806b15d3740ce6ace2fced9ef9d3b/torchvision/models/maxvit.py#L568) in JAX using the Flax module, we will implement the following required modules:

- `MaxViT`
  - `MaxVitBlock`
    - `MaxVitLayer`
      - `MBConv`
        - `Conv2dNormActivation`
        - `SqueezeExcitation`
      - `PartitionAttentionLayer`
        - `RelativePositionalMultiHeadAttention`
        - `WindowDepartition`
        - `WindowPartition`
        - `SwapAxes`
        - `StochasticDepth`
```

File: docs/JAX_porting_PyTorch_model.md (hunk 9 above)
Comment (vfdev-5) [reply docs/JAX_porting_PyTorch_model.md:1355]: No, inference mode it is a term meaning that we do not track grads on the forward pass and do not update buffers.

File: docs/JAX_porting_PyTorch_model.md (hunk 18 above)
Comment (vfdev-5) [reply docs/JAX_porting_PyTorch_model.md:141]: yes, this is too magic.

File: docs/JAX_porting_PyTorch_model.md (hunk 1 above)
Comment (melissawm) [reply docs/JAX_porting_PyTorch_model.md:131]: Ah that does help!

File: docs/JAX_porting_PyTorch_model.md (hunk 17 above)
Comment (jakevdp) [reply docs/JAX_porting_PyTorch_model.md:139]: Then maybe something like this?
```suggestion
- Pooling layers like `nn.AvgPool2d(...)` -> `lambda x: nnx.avg_pool(x, ...)`
- ```

**Comment by vfdev-5** [comment]
@melissawm @jakevdp thanks for the review, I applied majority of your suggestions to this tutorial

**Review by melissawm — APPROVED** [review]
LGTM!

**Comment by melissawm** [comment]
The errors on ReadtheDocs are about the headings: 
```
/home/docs/checkouts/readthedocs.org/user_builds/jax-ai-stack/checkouts/71/docs/JAX_porting_PyTorch_model.ipynb.rst:180002: WARNING: Non-consecutive header level increase; H2 to H4 [myst.header]
```

You might need to change them to be H3 instead.

**Review by jakevdp — APPROVED** [review]
